aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -101,5 +102,115 @@ Tensor nested_from_padded_cuda(
     return at::native::nested_from_padded_generic(padded, sizes);
   }
 }
+
+Tensor _collapse_two_dims_3(Tensor input, int64_t dim1, int64_t dim2) {
+  TORCH_CHECK(dim1 > 0, "dim1: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 > 0, "dim2: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 - 1 == dim1, "dim2 must be one more than dim1.")
+  TORCH_CHECK(dim1 == 1, "dim1 must be 1.")
+  TORCH_CHECK(input.dim() == 3, "Expected input to be 3 dim.");
+
+  auto* nt_input = get_nested_tensor_impl(input);
+  TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+  Tensor nt_sizes = nt_input->get_nested_size_tensor();
+  const auto& input_buffer = nt_input->get_buffer();
+
+  Tensor sizes_dim1 = at::native::narrow(nt_sizes, 1, 0, 1);
+  Tensor sizes_dim2 = at::native::narrow(nt_sizes, 1, 1, 1);
+
+  Tensor new_nt_sizes;
+  if (dim1 == 1) {
+    Tensor collapsed_sizes = sizes_dim1 * sizes_dim2;
+    new_nt_sizes = collapsed_sizes.contiguous();
+  }
+  Tensor result = at::detail::make_tensor<NestedTensorImpl>(nt_input->get_buffer(), new_nt_sizes);
+  TORCH_CHECK(result.dim() == 2, "Expected result to be 2 dimensional.");
+  return result;
+}
+
+Tensor batch_offsets_from_efficient_size(Tensor ef_sizes) {
+  int64_t* nt_sizes_ptr = ef_sizes.data_ptr<int64_t>();
+  Tensor offsets = at::empty({1 + ef_sizes.size(0)}, at::kLong);
+  int64_t* offsets_ptr = offsets.data_ptr<int64_t>();
+  offsets_ptr[0] = 0;
+  int64_t ef_sizes_size_1 = ef_sizes.size(1);
+  for (int64_t i = 0; i < ef_sizes.size(0); i++) {
```

Comment:
cache size(0) too, make the locals const, probably have to use c10::irange since we're pytorch core here and somebody will come along and change it if we don't